File: 1121-partition-array-for-maximum-sum/partition-array-for-maximum-sum.py
class Solution:
    def maxSumAfterPartitioning(self, arr: List[int], k: int) -> int:
        memo = {}

        def helper(idx):
            if idx in memo:
                return memo[idx]

            if idx >= len(arr):
                return 0
            
            curMax = 0
            val = 0

            for i in range(idx, min(len(arr), idx + k)):
                curMax = max(curMax, arr[i])
                val = max(val, curMax * (i - idx + 1) + helper(i + 1))
            
            memo[idx] = val
            return memo[idx]
        
        return helper(0)

File: 1121-partition-array-for-maximum-sum/partition-array-for-maximum-sum.js
/**
 * @param {number[]} arr
 * @param {number} k
 * @return {number}
 */
var maxSumAfterPartitioning = function(arr, k) {
    const memo = {};

    const helper = (index) => {
        if (index in memo) {
            return memo[index];
        }

        if(index >= arr.length) {
            return 0
        }

        let subarrMax = 0
        let subArrVal = 0

        for (let i = index; i<Math.min(arr.length, index + k); i++) {
            subarrMax = Math.max(subarrMax, arr[i]);
            subArrVal = Math.max(subArrVal, subarrMax * (i - index + 1) + helper(i + 1));
        }

        memo[index] = subArrVal;
        return memo[index];
    }

    return helper(0);
};